design/0001-frr.md

```diff
@@ -0,0 +1,347 @@
+* [BFD support](https://github.com/metallb/metallb/issues/396)
+* [BGP Failover too slow](https://github.com/metallb/metallb/issues/485)
+
+FRR is a mature Linux Foundation routing protocol suite based on Quagga that has
+been used in many production deployments. As such, it has been proven in terms
+of its maturity, flexibility (as can be seen by the [broad range of features](http://docs.frrouting.org/en/latest/bgp.html)
+it supports), scalability, security, reliability and performance. It also
+provides detailed logging features to aid debugging.
+
+### Goals
+
+* Provide a configuration option to allow use of FRR as a BGP agent in a MetalLB
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+
+### Non-Goals
+
+* Coexistence of two types of BGP agents (e.g. MetalLB native BGP agent AND FRR
+  BGP agent) is not a goal of this work
+* The purpose of this is not to provide an interface to plug any BGP agent into
+  MetalLB. Although, this may provide a useful template for others to do this.
+* This will not support runtime changing of BGP agent type.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a developer, I want to be able to run a comprehensive set of BGP system
+tests, preferably based on kubernetes e2e tests, in order to ensure the MetalLB
+BGP implementation works as expected.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to specify a configuration option
+when deploying MetalLB in order to select the MetalLB BGP agent to be used by
+MetalLB
+
+#### Story 3
+
+As a developer, I want to be able to configure FRR from MetalLB using FRR Go
+bindings in order to improve performance, reliability and debugging
+capabilities.
+
+### Risks and Mitigations
+
+FRR does not provide documented northbound Go bindings to allow configuration of
+BGP. There is an experimental gRPC interface. This interface may need to be
+productionized through the FRR community. When this has been satisfactorily
+achieved, we can start [Story 3](#story-3). Until that time, in order to
+mitigate this risk, we can wrap the FRR ‘vtysh’ command line interface in Go
+functions in order to configure FRR.
+
+## Design Details
+
+### Current Design
+
+Metallb’s speaker is deployed as a container with a Dockerfile specified
+[here](https://github.com/metallb/metallb/blob/main/speaker/Dockerfile) and
+running the [“speaker” Golang application](https://github.com/metallb/metallb/tree/main/speaker).
+This "speaker" application is the main application that manages advertisement of
+prefixes based on the creation of kubernetes services. The docker image and the
+golang executable get built through [tasks.py](https://github.com/metallb/metallb/blob/main/tasks.py)
+as part of the Metallb build process.
+
+The "speaker" application supports a Layer 2 mode and a BGP mode. The following
+discussion focuses on BGP mode.
+
```

Comment:
do we need to have debugging/troubleshooting section and explain how can we use FRR BGP debugging ?

Reply:
I don't think it is needed for this enhancement but I think it should be added to the implementation. I will try to add something to the enhancement to remind us to add that documentation to the implementation.

Reply:
I added this to the Goals section.